jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVNode.java

```diff
@@ -0,0 +1,555 @@
+        return "BlossomVNode pos = " + pos + ", dual: " + dual + ", true dual: " + getTrueDual()
+                + ", label: " + label + (isMarked ? ", marked" : "") + (isProcessed ? ", processed" : "")
+                + (blossomParent == null || isOuter ? "" : ", blossomParent = " + blossomParent.pos) +
+                (matched == null ? "" : ", matched = " + matched);
+    }
+
+    /**
+     * Represents nodes' labels
+     */
+    public enum Label {
+        /**
+         * The node is on an even layer in the tree (root has layer 0)
+         */
+        PLUS,
+        /**
+         * The node is on an odd layer in the tree (root has layer 0)
+         */
+        MINUS,
+        /**
+         * This node doesn't belong to any tree and is matched to some other node
+         */
+        INFINITY
+    }
+
+    /**
+     * An iterator for traversing the edges incident to this node.
+     * <p>
+     * This iterator has a feature that during every step it knows the next edge it'll return to the caller.
+     * That's why it is safe to modify the current edge (move it to another node, for example).
+     */
+    public class IncidentEdgeIterator implements Iterator<BlossomVEdge> {
+
+        /**
+         * The direction of the current edge
+         */
+        private int currentDir;
+        /**
+         * Direction of the {@code nextEdge}
+         */
+        private int nextDir;
+        /**
+         * The edge that will be returned after the next call to {@link IncidentEdgeIterator#next()}.
+         * Is null if all incident edges of the current node have been traversed.
+         */
+        private BlossomVEdge nextEdge;
+
+        /**
+         * Constructs a new instance of the IncidentEdgeIterator.
+         */
+        public IncidentEdgeIterator() {
+            if (first[0] == null) {
+                nextEdge = first[1];
+                nextDir = 1;
+
+            } else {
+                nextEdge = first[0];
+                nextDir = 0;
+            }
+        }
+
+        /**
+         * Returns the direction of the edge returned by this iterator
+         *
+         * @return the direction of the edge returned by this iterator
+         */
+        public int getDir() {
+            return currentDir;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return nextEdge != null;
+        }
+
+        @Override
+        public BlossomVEdge next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            BlossomVEdge result = nextEdge;
+            advance();
+            return result;
+        }
+
+        /**
+         * Advances this iterator to the next incident edge. If previous edge was the last one with direction
+         * 0, then the direction of this iterator changes. If previous edge was the last incident edge, then
+         * {@code nextEdge} becomes null.
+         */
+        private void advance() {
```

Comment:
We can't get rid of the currentDir variable and, therefore, avoid assigning nextDir to it. But that will slow down getDir() method. Since this iterator simply traverses all incident edges of a node, it can't be significantly optimized.